File: python/railroaded/tables/agencies.py
from __future__ import annotations

import os
from typing import Optional

import seared as s

from ..models import Agency
from ..util import load_list


@s.seared
class Agencies(s.Seared):
    '''
    Serializable dataclass table mapping `str` IDs to `Agency` records.

    Attributes:
        agencies (list[Agency]):
            a `list` of all `Agency` records in the `Agencies` table
        data (dict[str, Agency]):
            a `dict` mapping `str` IDs to `Agency` records
        ids (list[str]):
            a `list` of all `str` IDs in the `Agencies` table
    '''

    ### ATTRIBUTES ###
    # Required
    data: dict[str, Agency] = s.T(
        schema=Agency.SCHEMA,
        keyed=True,
        required=True
    )
    '''a `dict` mapping `str` IDs to `Agency` records'''


    ### CLASS METHODS ###
    @classmethod
    def from_gtfs (cls, path: str) -> Agencies:
        '''
        Returns an `Agencies` table populated from the GTFS data at `path`.

        Parameters:
            path (str):
                the path to the GTFS dataset

        Returns:
            agencies (Agencies):
                an `Agencies` table populated from the GTFS data at `path`
        '''
        agencies: list[Agency] = load_list(
            path = os.path.join(path, 'agency.txt'),
            schema = Agency.SCHEMA
        )
        return Agencies({ a.id: a for a in agencies })


    ### PROPERTIES ###
    @property
    def agencies (self) -> list[Agency]:
        '''a `list` of all `Agency` records in the `Agencies` table'''
        return list(self.data.values())
    
    @property
    def ids (self) -> list[str]:
        '''a `list` of all `str` IDs in the `Agencies` table'''
        return list(self.data.keys())
    

    ### MAGIC METHODS ###
    def __getitem__ (self, id: str) -> Optional[Agency]:
        '''
        Returns the `Agency` record associated with the `id` if it exists,
        otherwise returns `None`.
        
        Parameters:
            id (str):
                the `str` id associated with the `Agency` record to retrieve

        Returns:
            record (Optional[Agency]):
                the `Agency` record associated with `id` if it exists, otherwise 
                `None`
        '''
        return self.data.get(id, None)

File: typescript/src/tables/agencies.ts
import type { GTFSAgency, MGTFSAgency } from '../models'
import { Agency } from '../models'
import { loadList } from '../util'


export type MGTFSAgencies = {
    data: Record<string, MGTFSAgency>
}


/**
 * Utility class for loading and accessing a table mapping `string` IDs to
 * `Agency` records.
 * 
 * @property agencies - a `List` of `Agency` records in the `Agencies` table
 * @property data - a `Record` mapping `string` IDs to `Agency` records
 * @property ids - a `List` of all `string` IDs in the `Agencies` table
 * 
 * @see Agency
 */
export default class Agencies {
    
    agencies: Agency[]
    data: Record<string, Agency>
    ids: string[]

    constructor (data: MGTFSAgencies) {
        this.data = Object.keys(data.data).reduce(
            (prev, curr) => {
                prev[curr] = new Agency(data.data[curr]!)
                return prev
            },
            {} as Record<string, Agency>
        )
        this.agencies = Object.values(this.data)
        this.ids = Object.keys(this.data)
    }

    
    static async fromGTFS (
                data: Record<string, string>
            ): Promise<MGTFSAgencies> {
        const agencies = await loadList<GTFSAgency>(data['agency.txt']!)
            .then(gtfsAgencies => gtfsAgencies.map(
                agency => Agency.fromGTFS(agency)
            ))
        return {
            data: agencies.reduce(
                (prev, curr) => {
                    prev[curr.id] = curr
                    return prev
                },
                {} as Record<string, MGTFSAgency>
            )
        }
    }

    toGTFS (): MGTFSAgencies {
        return {
            data: Object.keys(this.data).reduce(
                (prev, curr) => {
                    prev[curr] = this.data[curr]!.toMGTFS()
                    return prev
                },
                {} as Record<string, MGTFSAgency>
            )
        }
    }
    

    get (id: string): Agency | undefined {
        return this.data[id]
    }
}